### ISSUE TITLE
Malformed oauth URL

### ISSUE THREAD

**Issue by dpfavand** [classy-org/classy-node#102]
We have an app that was working fine until last week. Suddenly the app was not able to get an auth token from the Classy API.

The exact error message from classy-node is `Error: "{\"request_path\":\"\\/\\/oauth2\\/auth\",\"message\":\"API not found with these values\",\"request_host\":[\"api.classy.org\"]}\n"`

Note the two slashes before oauth2 (`//oauth2/auth`). The library had been generating the malformed URL for ages but the API server was accepting it until sometime recently.

I believe this is caused by joining an empty string in the path if isAuthRequest is true. https://github.com/classy-org/classy-node/blob/master/src/ClassyResource/_createFullPath.js#L18

**Comment by mjalkio** [comment]
I don't believe the issue is with the empty string. `path.join('', 'hello', '', 'world')` returns `'hello/world'`. Is it possible that you included a trailing slash in the `baseUrl` when you instantiate your classy-node instance?

**Comment by dpfavand** [comment]
ah, well it must be something else then. The error occurs on the following minimal reproduction:

```
const Classy = require("classy-node");

const classy = new Classy({
  clientId: "XXXXXXXXXXXX",
  clientSecret: "XXXXXXXXXXXX",
  requestDebug: true
});

classy.app();
```

**Comment by brownjava** [comment]
What version of classy-node are you using?

**Comment by dpfavand** [comment]
2.6.2

**Comment by brownjava** [comment]
Huh.  Doesn't repro for me with 2.6.2.  The URI is correct:

`> echo "const Classy = require(\"classy-node\");  const classy = new Classy({   clientId: \"XXXXXXXXXXXX\",   clientSecret: \"XXXXXXXXXXXX\",   requestDebug: true });  classy.app();" > test.js ; and node test.js
{ debugId: 1,
  uri: 'https://api.classy.org/oauth2/auth',
  method: 'POST',
  headers:
   { Accept: 'application/json',
     'Content-Type': 'application/x-www-form-urlencoded',
     host: 'api.classy.org',
     'content-length': 79 },
  body: 'client_id=XXXXXXXXXXXX&client_secret=XXXXXXXXXXXX&grant_type=client_credentials' }
{ debugId: 1,
  headers:
   { date: 'Wed, 01 Aug 2018 19:08:23 GMT',
     'content-type': 'application/json; charset=utf-8',
     'content-length': '79',
     connection: 'close',
     'cache-control': 'no-cache',
     server: 'nginx',
     'x-kong-limit': 'time-consumed=49',
     'x-kong-upstream-latency': '58',
     'x-kong-proxy-latency': '19',
     via: 'kong/0.9.3' },
  statusCode: 400,
  body: '{"error_description":"Invalid client authentication","error":"invalid_client"}\n' }
App token failure
(node:65364) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: "{\"error_description\":\"Invalid client authentication\",\"error\":\"invalid_client\"}\n"
(node:65364) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.`

**Comment by dpfavand** [comment]
ok it must only occur on Windows. Just ran the same code in WSL/Ubuntu and it works. The server is an Azure app service running Windows.

That said, path.join('', 'hello', '', 'world') on Windows does generate 'hello\world' so I agree that's probably not the source of the issue.

**Comment by brownjava** [comment]
I can confirm.  This repros on Windows.

**Comment by brownjava** [comment]
OK, so here's the problem.  On line 22 of _createFullPath.js we call path.normalize('/oauth2/auth').  However, path.normalize is a *filesystem* path normalizer and gives different results depending on the OS on which it is running.  On UNIX OSes it results in '/oauth2/auth', and on Windows it results in '\\oauth2\\auth'.

**Comment by brownjava** [comment]
Pull request: https://github.com/classy-org/classy-node/pull/103

**Comment by mlingner** [comment]
#103 has now been merged and is included in v2.6.3. Thanks @brownjava and @dpfavand.
